Added lfw dataset

SinghKislay opened the pull request:
Ran on a smaller amount of images as there are a lot of possibilty of triplets from 13,000 images.
Dowload and extract worked just fine. 

tfds.core.DatasetInfo(
    name='lfw',
    version=2.0.0,
    description='Labeled Faces in the Wild: A Database for Studying Face Recognition in Unconstrained Environments',
    urls=['http://vis-www.cs.umass.edu/lfw/#resources'],
    features=FeaturesDict({
        'anchor': Image(shape=(250, 250, 3), dtype=tf.uint8),
        'negative': Image(shape=(250, 250, 3), dtype=tf.uint8),
        'positive': Image(shape=(250, 250, 3), dtype=tf.uint8)
    },
    total_num_examples=69345,
    splits={
        'train': <tfds.core.SplitInfo num_examples=69345>
    },
    supervised_keys=('', ''),
    citation='"""
        @TechReport{LFWTech,
          author = {Gary B. Huang and Manu Ramesh and Tamara Berg and
                          Erik Learned-Miller},
          title = {Labeled Faces in the Wild: A Database for Studying
                          Face Recognition in Unconstrained Environments},
          institution = {University of Massachusetts, Amherst},
          year = 2007,
          number = {07-49},
          month = {October}}
    """',
)

SinghKislay commented:
I signed it! the user agreement.

SinghKislay commented:
@us could you please review the code. Thanks.

us commented on tensorflow_datasets/image/lfw.py:
```diff
@@ -0,0 +1,105 @@
+            urls=[self.URL],
+            citation=LFW_CITATION,
+        )
+    
```
You should run `pylint` to cleanup unnecessary empty lines. You can check this [link](https://www.tensorflow.org/community/contribute/code_style) about code style.

us commented on tensorflow_datasets/image/lfw.py:
```diff
@@ -0,0 +1,105 @@
+            temp1=os.listdir(lst_path)
+            if(len(temp1)>1):
+                lfw_mod.append(lst)
+        #print(len(lfw_mod))
```
I think you forget to delete this comment line.

us replied:
one more forgotten line

us commented on tensorflow_datasets/image/lfw.py:
```diff
@@ -0,0 +1,105 @@
+                                    else:
+                                        for __i in total_images_negative:
+                                                triplet_dict={'anchor':os.path.join(path,path_list[img_no]),'positive':os.path.join(path,path_list[img_no+offset]),'negative':os.path.join(path_list_negative,__i)}
+                                                #print(triplet_dict)
```
one more forgotten line

us commented on tensorflow_datasets/image/lfw.py:
```diff
@@ -0,0 +1,105 @@
+
+_URL="http://vis-www.cs.umass.edu/lfw/lfw.tgz"
+LFW_IMAGE_SHAPE=(250,250,3)
+LFW_CITATION="""@TechReport{LFWTech,
```
You can edit your line like others for better readability.
Use `LFW_CITATION="""\`

us commented:
Thank you for contributing. 😊

SinghKislay commented:
these images are from the dataset itself. So i don't think they are copyrighted. I'll make the changes regarding the code style asap.

us commented:
I haven't see any license about this dataset. Please check this [link](https://academia.stackexchange.com/a/63157).

SinghKislay commented:
@us thank you for your feed-backs. I have changed the images to random generated noise(So, no license issue I hope :) ).  Please do tell if any other changes are required.

us commented:
Thank you :) But I think you forgot to change the code style.

SinghKislay commented:
I am sorry but I am a little lost here. I looked at other examples such as catsvsdogs.py and mnist.py. I added the indentations and spaces just like they did and also changed the citation as you mentioned. Could you please point out a little more, Thanks(I am a newbie to open source).

us commented:
Okay firstly you should check the [pylint](https://www.pylint.org/). It's simply checks your code styles,  detects your errors and rates your code. Then you should read [TensorFlow code style guide](https://www.tensorflow.org/community/contribute/code_style).
 In brief detect errors with pylint and fixed them.

SinghKislay commented:
@us thank-you for your patience. I have got a pylint score of 4.51(initially it was in negative) also I have followed the Tensorflow guideline of 2 spaces instead of 4. I think it should pass this time.

us commented:
@SinghKislay but 4.51/10 is a bad score. You should increase the score by checking errors. And I think you understood wrongly. Indentation should be 2 not 4. You are still use 4 space.

SinghKislay commented:
@us I got a little confused. Now I have got a score of 8.6. And the only errors are too many branches(it is in the helper function, nothing can be done about this) and too few public method and np doc-string for methods(No one added these so I left them as well).

us commented on tensorflow_datasets/image/lfw.py:
```diff
@@ -40,81 +40,83 @@
-              "negative": tfds.features.Image(shape=LFW_IMAGE_SHAPE),
+  """LFW Class"""
+  URL = "http://vis-www.cs.umass.edu/lfw/#resources"
+  VERSION = tfds.core.Version("2.0.0")
```
Why do you start from version 2.0.0 ?
Learn more about versioning by checking other datasets commits.

SinghKislay replied:
Changed the version to 0.1.0

us commented:
Maybe you can use `# pylint: disable=missing-docstring` to disable the warnings.

us commented on tensorflow_datasets/image/lfw.py:
```diff
@@ -0,0 +1,122 @@
+  #This is a helper function for making all possible triplets for siamese network(eg. FaceNet)
+  def triplet_maker(self, _path):
+    triplet_list = []
+    lfw = os.listdir(_path)
```
You shuld use `tf.io.gfile` for all filesystem access. Check this [link](https://stackoverflow.com/a/42936969/8401425) about why using that

SinghKislay replied:
I understand I have removed them with tf.gfile.ListDirectory

SinghKislay replied:
Also ran and checked, every thing seems to work fine on my system.

cyfra commented:
@us - thanks a lot for doing the code review. Appreciated.

@SinghKislay - seems that you're returning the data in the triplet form for Siamese Networks. In tfds we should try to return the dataset in as "raw" format as possible - in this case, it would be just images, as some people might want to use it in other research, where they don't require triplets.
And if needed, people can generate triplets themselves based on the raw image feed.

SinghKislay commented:
Hello @cyfra I have resolved the issue that you presented. Now the user only gets the image with name of the file. I ran and checked the new code on my system it seems to work. I have included a gist for the dataset_info file here: https://gist.github.com/SinghKislay/e42daac722192c59ddb7f9440702ae64
looking forward to hearing back from you.

us commented on tensorflow_datasets/image/__init__.py:
```diff
@@ -43,4 +43,6 @@
 from tensorflow_datasets.image.svhn import SvhnCropped
+from tensorflow_datasets.image.voc import Voc2007
 from tensorflow_datasets.image.lfw import LFW
+
```
empty line

us commented:
when you do the last changes, maybe you should open new pr that include simple commits. Because you have too much commits to single dataset.

us commented on tensorflow_datasets/testing/test_data/fake_examples/lfw/lfw/ReadME.txt:
```diff
@@ -0,0 +1,5 @@
+
+These are structured in just the way it was in original dataset
+
+imgae size= 250 X 250
```
image

SinghKislay commented:
@us, do u think everything is okay now.

us commented on tensorflow_datasets/image/lfw.py:
```diff
@@ -0,0 +1,95 @@
+
+class LFW(tfds.core.GeneratorBasedBuilder):
+  """LFW Class"""
+  URL = "http://vis-www.cs.umass.edu/lfw/#resources"
```
delete `#resources` in the end.

us commented on tensorflow_datasets/image/lfw.py:
```diff
@@ -0,0 +1,95 @@
+        description=("Labeled Faces in the Wild:\
+  A Database for Studying Face Recognition in Unconstrained Environments"),
+        features=tfds.features.FeaturesDict({
+            "image_name": tfds.features.Text(),
```
`label` could be better thane `image_name` to consistency and meaning.

SinghKislay replied:
@us there is no label here in this dataset. it's just dir by the name of a person in which there are pictures of the person, and hence the most logical thing would be to give the name of the image and the image itself. But if you have something else in mind please do suggest.

us commented:
I think yes. After checking my last reviews, you can close this pr and open a clean new pr.

SinghKislay commented:
Cool
